src/ir/module-utils.h

```diff
@@ -445,13 +445,21 @@ collectSignatures(Module& wasm,
 
   // Collect all the counts.
   Counts counts;
+  auto maybeAdd = [&](Type type) {
+    if (type.isRef()) {
+      auto heapType = type.getHeapType();
+      if (heapType.isSignature()) {
+        counts[heapType.getSignature()]++;
+      }
+    }
+  };
```

Comment:
Good idea to share that method, will refactor.

I don't think that method needs to handle tuple types, though? That's handled explicitly in visitExpression which is the one place where it is relevant, I believe (and we need to check if the expression is a control flow structure first there).

Reply:
If block types contain tuples, shouldn't we check each of the tuple element for heap signature too?

Reply:
I think that is what the code on line 470+ does. We would have the block type there, and then recursively traverse its elements, and then we'd get the signature types in there.

Reply:
Ah, I think you're right. Then I guess we don't need the separate tuple handling for `vars`(locals)?

Reply:
I added a testcase for this now. Does that match the problem you were thinking about?

Reply:
I think we do need the tuple handling there. The code lower down does recursive analysis of signatures - not types.

But thinking about it now, I am not sure why this is different from the other tuple handling code in `visitExpression`. I think they both achieve the same result - getting a signature from a type, which is then analyzed later - but one also creates a signature (with no inputs, and with those outputs) that is counted. I *think* that is because in the binary format, such a thing appears for control flow structures as `visitExpression` handles, but for a tuple in a local, we actually lower those down anyhow, they are not in the wasm binary format, so we don't need to count them (just to note their children). So I think it's correct, but I'm not as familiar with the tuple stuff as @tlively ?

Reply:
Yes I think so, thanks!

Reply:
Yes, that all sounds correct.